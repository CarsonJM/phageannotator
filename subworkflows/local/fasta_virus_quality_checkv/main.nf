//
// Assess virus quality with Checkv
//
include { CHECKV_DOWNLOADDATABASE   } from '../../../modules/nf-core/checkv/downloaddatabase/main'
include { CHECKV_ENDTOEND           } from '../../../modules/nf-core/checkv/endtoend/main'
include { QUALITYFILTERVIRUSES      } from '../../../modules/local/qualityfilterviruses/main'

workflow FASTA_VIRUS_QUALITY_CHECKV {
    take:
    virus_fasta_gz  // [ [ meta ], fasta.gz ]   , assemblies/genomes (mandatory)
    checkv_db       // [ checkv_db ]            , CheckV database directory (optional)

    main:
    ch_versions = Channel.empty()

    // if checkv DB exists, skip CHECKV_DOWNLOADDATABASE
    if ( checkv_db ){
        ch_checkv_db = checkv_db
    } else {
        //
        // MODULE: download standard CheckV database
        //
        ch_checkv_db    = CHECKV_DOWNLOADDATABASE( ).checkv_db
        ch_versions     = ch_versions.mix( CHECKV_DOWNLOADDATABASE.out.versions )
    }

    //
    // MODULE: Run CheckV end to end
    //
    ch_quality_summary_tsv  = CHECKV_ENDTOEND ( virus_fasta_gz, ch_checkv_db.collect() )
    ch_viruses_fna_gz       = CHECKV_ENDTOEND.out.viruses
    ch_proviruses_fna_gz    = CHECKV_ENDTOEND.out.proviruses
    ch_versions             = ch_versions.mix( CHECKV_ENDTOEND.out.versions )

    // create channel for input into QUALITY_FILTER_VIRUSES
    ch_quality_filter_viruses_input = ch_viruses_fna_gz
        .join( ch_proviruses_fna_gz )
        .join( ch_quality_summary_tsv )
        .multiMap { it ->
            viruses: [ it[0], it[1] ]
            proviruses: [ it[0], it[2] ]
            quality_summary: [ it[0], it[3] ]
        }

    //
    // MODULE: Quality filter viruses
    //
    ch_filtered_viruses_fna_gz = QUALITYFILTERVIRUSES (
        ch_quality_filter_viruses_input.viruses,
        ch_quality_filter_viruses_input.proviruses,
        ch_quality_filter_viruses_input.quality_summary
    ).filtered_viruses
    ch_versions = ch_versions.mix( QUALITYFILTERVIRUSES.out.versions )

    emit:
    filtered_viruses_fna_gz     = ch_quality_filtered_viruses_fna_gz    // [ [ meta ], viruses.fna.gz ]       , FASTA file containing quality-filtered viruses
    quality_summary_tsv         = ch_quality_summary_tsv                // [ [ meta ], quality_summary.tsv ]  , TSV file containing quality data
    versions                    = ch_versions                           // [ versions.yml ]
}
