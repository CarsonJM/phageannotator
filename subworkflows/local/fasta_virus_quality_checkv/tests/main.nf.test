nextflow_workflow {

    name "Test Subworkflow: FASTA_VIRUS_QUALITY_CHECKV"
    script "subworkflows/local/fasta_virus_quality_checkv/main.nf"
    workflow "FASTA_VIRUS_QUALITY_CHECKV"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fasta_virus_quality_checkv"
    tag "fasta_virus_quality_checkv_full"


    test("Full subworkflow") {

        when {
            params {
                outdir              = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], [ file('https://github.com/nf-core/test-datasets/blob/modules/data/genomics/sarscov2/genome/genome.fasta.gz', checkIfExists: true) ] ])
                input[1] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.trace.succeeded().size() == 3 },
                { assert snapshot(path("$outputDir/VirusQuality").list()).match("fasta_virus_quality_checkv") }
            )
        }
    }
}
