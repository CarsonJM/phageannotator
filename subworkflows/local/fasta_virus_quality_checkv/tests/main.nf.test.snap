{
    "Parameters: default": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "viruses.fna.gz:md5,1b6b8c20b3e1b5aa7dafccf01529c3be"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "proviruses.fna.gz:md5,8e31b56780f8b1e604fc8dfa5c4b5921"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ]
                ],
                "3": [
                    "versions.yml:md5,58c018640348d4755b08dd850a477465",
                    "versions.yml:md5,e4d4311cae4bcb6711e82eef25b187ff",
                    "versions.yml:md5,bde8142f045061369c16cd56d51698ed"
                ],
                "proviruses_fna_gz": [
                    [
                        {
                            "id": "test"
                        },
                        "proviruses.fna.gz:md5,8e31b56780f8b1e604fc8dfa5c4b5921"
                    ]
                ],
                "quality_summary_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,58c018640348d4755b08dd850a477465",
                    "versions.yml:md5,e4d4311cae4bcb6711e82eef25b187ff",
                    "versions.yml:md5,bde8142f045061369c16cd56d51698ed"
                ],
                "viruses_fna_gz": [
                    [
                        {
                            "id": "test"
                        },
                        "viruses.fna.gz:md5,1b6b8c20b3e1b5aa7dafccf01529c3be"
                    ]
                ]
            }
        ],
        "timestamp": "2023-10-02T18:18:29+0000"
    }
}